File: src/Scripting/Ruby/patterns/null_object.rb
# frozen_string_literal: true

module RubyPatterns
  module NullObject
    module_function

    def run
      null_logger = Object.new
      null_logger.define_singleton_method(:log) { |_message| nil }
      service = Struct.new(:logger) do
        def run
          logger.log('run')
          :ok
        end
      end
      raise 'Null Object failed' unless service.new(null_logger).run == :ok
      true
    end
  end
end

RubyPatterns::NullObject.run if $PROGRAM_NAME == __FILE__

File: src/Scripting/PHP/patterns/null_object.php
<?php declare(strict_types=1);
$check = static fn(bool $ok) => $ok ?: throw new RuntimeException('null_object');
$nullLogger = new class { public function log(string $message): void {} };
$service = static function(object $logger): string { $logger->log('work'); return 'done'; };
$check($service($nullLogger) === 'done');
